src/frequenz/sdk/timeseries/_moving_window.py

```diff
@@ -0,0 +1,182 @@
+            raise
+
+    async def _run_impl(self) -> None:
+        """Awaits samples from the receiver and updates the underlying ringbuffer."""
+        async for sample in self._resampled_data_recv:
+            log.debug("Received new sample: %s", sample)
+            log.debug("Updating ring buffer with latest sample")
+
+            # missing values coming from the logical meter are marked as NaN
+            missing = False
+            assert isinstance(sample.value, float)
+            if isnan(sample.value):
+                missing = True
+
+            self._buffer.update(sample.timestamp, sample.value, missing)
+
+        log.error("Channel has been closed")
+
+    def __len__(self) -> int:
+        """
+        Return the size of the `MovingWindow`s underlying buffer.
+
+        Returns:
+            The size of the `MovingWindow`.
+        """
+        return len(self._buffer)
+
+    def __getitem__(self, key: (int | datetime) | slice) -> float | ArrayLike:
+        """
+        Return a sub window of the `MovingWindow`.
+
+        The `MovingWindow` is accessed either by timestamp or by index
+        or by a slice of timestamps or integers.
+
+        * If the key is an integer, the float value of that key
+          at the given position is returned.
+        * If the key is a datetime object, the float value of that key
+          that corresponds to the timestamp is returned.
+        * If the key is a slice of timestamps or integers, an ndarray is returned,
+          where the bounds correspond to the slice bounds.
+          Note that a half open intervall, which is open at the end, is returned.
+
+        Args:
+            key: Either an integer or a timestam or a timestamps or integers.
+
+        Raises:
+            IndexError: when requesting an out of range timestamp or index
+            TypeError: when the key is not a datetime or slice object.
+
+        Returns:
+            Union(float, ArrayLike): A float of the key is a number or a timestamp.
+                An numpy array if the key is a slice.
+        """
+        if isinstance(key, slice):
+            log.debug("Returning slice for [%s:%s].", key.start, key.stop)
+            # we are doing runtime typechecks since there is no abstract slice type yet
+            # see also (https://peps.python.org/pep-0696)
+            if isinstance(key.start, datetime) and isinstance(key.stop, datetime):
+                return self._buffer.window(key.start, key.stop, self._copy_buffer)
+            if isinstance(key.start, int) and isinstance(key.stop, int):
+                window = self._buffer[key]
+                # we need to assert here to make the type checker happy
+                assert isinstance(window, np.ndarray)
+                return window
+        elif isinstance(key, datetime):
+            log.debug("Returning value at time %s ", key)
+            value = self._buffer[self._buffer.datetime_to_index(key)]
+            return value
```

Comment:
here you might want to return directly `self._buffer[self._buffer.datetime_to_index(key)]` instead to avoid declaring `value`